File: day05/part1.pl
#!/usr/bin/env perl
use List::Util qw(min max);
use strict;

my %CELLS;
while (<>) {
  /^(\d+),(\d+) -> (\d+),(\d+)$/;
  if ($1 == $3 || $2 == $4) {
    for my $x (min($1,$3)..max($1,$3)) {
      for my $y (min($2,$4)..max($2,$4)) {
        $CELLS{int($x)}{int($y)} += 1;
      }
    }
  }
}

my $count;
foreach my $x (keys %CELLS) {
  foreach my $y (keys %{%CELLS{$x}}) {
    if ($CELLS{$x}{$y} > 1) {
      $count += 1;
    }
  }
}
print "$count\n";

File: day09/part1.php
#!/usr/bin/env php
<?php

$CELLS = array();
while ($line = fgets(STDIN)) {
	$CELLS[] = str_split(trim($line));
}
$H = count($CELLS);
$W = count($CELLS[0]);

for ($x = 0; $x < $W; $x++) {
	for ($y = 0; $y < $H; $y++) {
		$CELLS[$y][$x] = intval($CELLS[$y][$x]);
	}
}

function value($x, $y) {
	global $H, $W, $CELLS;
	if ($x >= 0 && $x < $W && $y >= 0 && $y < $H) {
		return $CELLS[$y][$x];
	}
	return 9;
}

function neighbour_values($x, $y) {
	return [
		value($x+1, $y),
		value($x-1, $y),
		value($x, $y+1),
		value($x, $y-1),
	];
}

$risk = 0;
for ($x = 0; $x < $W; $x++) {
	for ($y = 0; $y < $H; $y++) {
		if ($CELLS[$y][$x] < min(neighbour_values($x, $y))) {
			$risk += $CELLS[$y][$x] + 1;
		}
	}
}

echo "$risk";
